source/common/common/thread.h

```diff
@@ -80,24 +83,28 @@ enum class AtomicPtrAllocMode { DoNotDelete, DeleteOnDestruct };
 
 // Manages an array of atomic pointers to T, providing a relatively
 // contention-free mechanism to lazily get a T* at an index, where the caller
-// provides a mechanism to instantiate a T* under lock, if one is not already
+// provides a mechanism to instantiate a T* under lock, if one has not already
 // been stored at that index.
 //
 // alloc_mode controls whether allocated T* entries should be deleted on
 // destruction of the array. This should be set to AtomicPtrAllocMode::DoNotDelete
 // if the T* returned from MakeObject are managed by the caller.
-template <class T, uint32_t size, AtomicPtrAllocMode alloc_mode> class AtomicPtrArray {
+template <class T, uint32_t size, AtomicPtrAllocMode alloc_mode>
+class AtomicPtrArray : NonCopyable {
 public:
   AtomicPtrArray() {
-    for (auto& ptr : data_) {
-      ptr = nullptr;
+    for (std::atomic<T*>& atomic_ref : data_) {
+      atomic_ref = nullptr;
     }
   }
 
   ~AtomicPtrArray() {
     if (alloc_mode == AtomicPtrAllocMode::DeleteOnDestruct) {
-      for (auto& ptr : data_) {
-        delete ptr;
+      for (std::atomic<T*>& atomic_ref : data_) {
+        T* ptr = atomic_ref.load();
+        if (ptr != nullptr) {
```

Comment:
This is not actually thread safe. Can you add more comments as to why this is OK based on usage or some other factor? Or should the mutex be held. Also, should data_ have lock annotations on it?

Reply:
By the time we are destructing the object, there better be only one thread accessing it or mutexes cannot help :)

https://stackoverflow.com/questions/646427/should-destructors-be-threadsafe

Reply:
OK sorry I had it in my head that we are taking a reference to a vector and using it, but this makes sense.